flax/core/scope.py

```diff
@@ -14,13 +14,52 @@
 
 """Flax functional core: Scopes.
 
-The goal of the Flax functional core is to provide a purely functional 
-abstraction that takes care of various types of bookkeeping such as parameter
-and RNG management. The main abstraction is Scope, which contains parameters
-and RNGs, and which can be nested.
-
-TODO: elaborate.
-
+Flax deep learning models contain variables, which can be of different 
+categories such as trainable parameters, or internal state variables that aren't
+parameters (e.g., batch statistics). For instance, the top-level variable
+dictionary of a Flax model using batch normalization may look as follows::
+  {
+    "params": {
+      "Conv1": { "weight": ..., "bias": ... },
+      "BatchNorm1": { "scale": ..., "mean": ... },
+      "Conv2": {...}
+    },
+    "batch_stats": {
+      "BatchNorm1": { "moving_mean": ..., "moving_average": ...}
+    }
+  }
+
+This variable dict has two so-called "collections": `params` (containing the 
+trainable parameters) and `batch_stats` (containing the --non-trainable-- batch
+statistics).
+
+The functional core Scope defined below allows easy access to all variables in a
+layer. For instance, from inside the BathcNorm1 layer, the variable "mean" is
+accessed as follows::
+  mean = scope.variable('batch_stats', 'moving_mean', jnp.zeros, (,))
+
+This will initialize it using `jnp.zeros` with shape `(,)`. The function 
+`scope.params` accesses variables of collection "params" directly::
+  weight = scope.params('scale', jnp.zeros, (3, 6))
+
+Scope are not constructed directly, but passed as a first argument to functions
+to which they should be applied. E.g., the scopes of the example above::
+  def conv(scope, x, ...)::
+    weight = scope.params('weight', ...)
+    ...
+
+  def module_fn(scope, x, ...):
+    conv = scope.child(conv, 'conv')(x, ...)
+    ...
+
+The function `init` is then used to initialize a Scope, which transforms it into
+a function that expects RNG keys for all collections::
```

Comment:
rng sequenced and collections aren't linked (except for paramaters).  But there is no 'batch_stats' rng